### ISSUE TITLE
AST merge of recursive function

### ISSUE THREAD

**Issue by balazske** [Ericsson/clang#282]
file `recursive.h`:
```
#pragma once
void recursive_function();
```

file `recursive.cpp`:
```
#include "recursive.h"
void recursive_function() {
	recursive_function();
}
```

file `root.cpp`:
```
#include "recursive.h"
void root() {
	recursive_function();
}

```
following commands produce the crash below:

`clang -cc1 -emit-pch -o recursive.ast recursive.cpp`
`clang -cc1 -ast-merge recursive.ast -fsyntax-only root.cpp`
```
Unsortable locations found
UNREACHABLE executed at llvm/tools/clang/lib/Basic/SourceManager.cpp:2057!
#0 0x00007f7753a64fd9 llvm::sys::PrintStackTrace(llvm::raw_ostream&) llvm/lib/Support/Unix/Signals.inc:398:0
#1 0x00007f7753a6506a PrintStackTraceSignalHandler(void*) llvm/lib/Support/Unix/Signals.inc:462:0
#2 0x00007f7753a634b0 llvm::sys::RunSignalHandlers() llvm/lib/Support/Signals.cpp:49:0
#3 0x00007f7753a64971 SignalHandler(int) llvm/lib/Support/Unix/Signals.inc:252:0
#4 0x00007f775037dcb0 (/lib/x86_64-linux-gnu/libc.so.6+0x36cb0)
#5 0x00007f775037dc37 gsignal /build/eglibc-ripdx6/eglibc-2.19/signal/../nptl/sysdeps/unix/sysv/linux/raise.c:56:0
#6 0x00007f7750381028 abort /build/eglibc-ripdx6/eglibc-2.19/stdlib/abort.c:91:0
#7 0x00007f77539a9331 bindingsErrorHandler(void*, std::string const&, bool) llvm/lib/Support/ErrorHandling.cpp:198:0
#8 0x00007f77528fd24e clang::SourceManager::isBeforeInTranslationUnit(clang::SourceLocation, clang::SourceLocation) const llvm/tools/clang/lib/Basic/SourceManager.cpp:2058:0
#9 0x00007f774924bd4e (anonymous namespace)::CompareOverloadCandidatesForDisplay::operator()(clang::OverloadCandidate const*, clang::OverloadCandidate const*) llvm/tools/clang/lib/Sema/SemaOverload.cpp:10340:0
#10 0x00007f774925dd2c void std::__insertion_sort<clang::OverloadCandidate**, (anonymous namespace)::CompareOverloadCandidatesForDisplay>(clang::OverloadCandidate**, clang::OverloadCandidate**, (anonymous namespace)::CompareOverloadCandidatesForDisplay) /usr/include/c++/4.8/bits/stl_algo.h:2159:0
#11 0x00007f774925cf16 void std::__final_insertion_sort<clang::OverloadCandidate**, (anonymous namespace)::CompareOverloadCandidatesForDisplay>(clang::OverloadCandidate**, clang::OverloadCandidate**, (anonymous namespace)::CompareOverloadCandidatesForDisplay) /usr/include/c++/4.8/bits/stl_algo.h:2232:0
#12 0x00007f774925c448 void std::sort<clang::OverloadCandidate**, (anonymous namespace)::CompareOverloadCandidatesForDisplay>(clang::OverloadCandidate**, clang::OverloadCandidate**, (anonymous namespace)::CompareOverloadCandidatesForDisplay) /usr/include/c++/4.8/bits/stl_algo.h:5502:0
#13 0x00007f774924c54d clang::OverloadCandidateSet::NoteCandidates(clang::Sema&, clang::OverloadCandidateDisplayKind, llvm::ArrayRef<clang::Expr*>, llvm::StringRef, clang::SourceLocation, llvm::function_ref<bool (clang::OverloadCandidate&)>) llvm/tools/clang/lib/Sema/SemaOverload.cpp:10452:0
#14 0x00007f7749251bfa FinishOverloadedCallExpr(clang::Sema&, clang::Scope*, clang::Expr*, clang::UnresolvedLookupExpr*, clang::SourceLocation, llvm::MutableArrayRef<clang::Expr*>, clang::SourceLocation, clang::Expr*, clang::OverloadCandidateSet*, clang::OverloadCandidate**, clang::OverloadingResult, bool) llvm/tools/clang/lib/Sema/SemaOverload.cpp:11787:0
#15 0x00007f77492520a6 clang::Sema::BuildOverloadedCallExpr(clang::Scope*, clang::Expr*, clang::UnresolvedLookupExpr*, clang::SourceLocation, llvm::MutableArrayRef<clang::Expr*>, clang::SourceLocation, clang::Expr*, bool, bool) llvm/tools/clang/lib/Sema/SemaOverload.cpp:11855:0
#16 0x00007f7748f5eb86 clang::Sema::ActOnCallExpr(clang::Scope*, clang::Expr*, clang::SourceLocation, llvm::MutableArrayRef<clang::Expr*>, clang::SourceLocation, clang::Expr*, bool) llvm/tools/clang/lib/Sema/SemaExpr.cpp:5190:0
#17 0x00007f7749f4568b clang::Parser::ParsePostfixExpressionSuffix(clang::ActionResult<clang::Expr*, true>) llvm/tools/clang/lib/Parse/ParseExpr.cpp:1664:0
#18 0x00007f7749f442d8 clang::Parser::ParseCastExpression(bool, bool, bool&, clang::Parser::TypeCastState, bool) llvm/tools/clang/lib/Parse/ParseExpr.cpp:1429:0
#19 0x00007f7749f3ff8a clang::Parser::ParseCastExpression(bool, bool, clang::Parser::TypeCastState, bool) llvm/tools/clang/lib/Parse/ParseExpr.cpp:525:0
#20 0x00007f7749f3ea88 clang::Parser::ParseAssignmentExpression(clang::Parser::TypeCastState) llvm/tools/clang/lib/Parse/ParseExpr.cpp:170:0
#21 0x00007f7749f3e880 clang::Parser::ParseExpression(clang::Parser::TypeCastState) llvm/tools/clang/lib/Parse/ParseExpr.cpp:120:0
#22 0x00007f7749f99d58 clang::Parser::ParseExprStatement() llvm/tools/clang/lib/Parse/ParseStmt.cpp:413:0
#23 0x00007f7749f9954d clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::AllowedConstructsKind, clang::SourceLocation*, clang::Parser::ParsedAttributesWithRange&) llvm/tools/clang/lib/Parse/ParseStmt.cpp:224:0
#24 0x00007f7749f98f06 clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::AllowedConstructsKind, clang::SourceLocation*) llvm/tools/clang/lib/Parse/ParseStmt.cpp:111:0
#25 0x00007f7749f9bfc9 clang::Parser::ParseCompoundStatementBody(bool) llvm/tools/clang/lib/Parse/ParseStmt.cpp:1001:0
#26 0x00007f7749f9f6f6 clang::Parser::ParseFunctionStatementBody(clang::Decl*, clang::Parser::ParseScope&) llvm/tools/clang/lib/Parse/ParseStmt.cpp:1967:0
#27 0x00007f7749fbd28b clang::Parser::ParseFunctionDefinition(clang::ParsingDeclarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::LateParsedAttrList*) llvm/tools/clang/lib/Parse/Parser.cpp:1212:0
#28 0x00007f7749ef7db1 clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, unsigned int, clang::SourceLocation*, clang::Parser::ForRangeInit*) llvm/tools/clang/lib/Parse/ParseDecl.cpp:1953:0
#29 0x00007f7749fbc288 clang::Parser::ParseDeclOrFunctionDefInternal(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec&, clang::AccessSpecifier) llvm/tools/clang/lib/Parse/Parser.cpp:980:0
#30 0x00007f7749fbc34d clang::Parser::ParseDeclarationOrFunctionDefinition(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*, clang::AccessSpecifier) llvm/tools/clang/lib/Parse/Parser.cpp:995:0
#31 0x00007f7749fbbaa4 clang::Parser::ParseExternalDeclaration(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*) llvm/tools/clang/lib/Parse/Parser.cpp:845:0
#32 0x00007f7749fbae49 clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&) llvm/tools/clang/lib/Parse/Parser.cpp:613:0
#33 0x00007f7749edef30 clang::ParseAST(clang::Sema&, bool, bool) llvm/tools/clang/lib/Parse/ParseAST.cpp:146:0
#34 0x00007f7751009866 clang::ASTFrontendAction::ExecuteAction() llvm/tools/clang/lib/Frontend/FrontendAction.cpp:1005:0
#35 0x00007f7750f3da29 clang::ASTMergeAction::ExecuteAction() llvm/tools/clang/lib/Frontend/ASTMerge.cpp:78:0
#36 0x00007f77510092d7 clang::FrontendAction::Execute() llvm/tools/clang/lib/Frontend/FrontendAction.cpp:906:0
#37 0x00007f7750f99361 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) llvm/tools/clang/lib/Frontend/CompilerInstance.cpp:981:0
#38 0x00007f7750a95d32 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) llvm/tools/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:251:0
#39 0x000000000044a4b8 (clang+0x44a4b8)
#40 0x0000000000440254 (clang+0x440254)
#41 0x0000000000440e52 (clang+0x440e52)
#42 0x00007f7750368f45 __libc_start_main /build/eglibc-ripdx6/eglibc-2.19/csu/libc-start.c:321:0
#43 0x000000000043dae9 (clang+0x43dae9)
Stack dump:
0.	Program arguments: clang -cc1 -ast-merge rf.ast -fsyntax-only root.cpp 
1.	root.cpp:3:21: current parser token ')'
2.	root.cpp:2:13: parsing function body 'root'
3.	root.cpp:2:13: in compound statement ('{}')
Aborted
```

**Comment by martong** [comment]
This crash only happens when the imported TU contains a function prototype.
There is no crash for the following example
`recursive.cpp`
```
//void f();
void f() {
	f();
}
```
`root.cpp`
```
void f();
void root() {
	f();
}
```

**Comment by martong** [comment]
Related pr #285

**Comment by dkrupp** [comment]
This patch brought in some unwanted asserts in xerces
https://github.com/Ericsson/clang/pull/299/files

please check why and resubmit the patch.

**Comment by martong** [comment]
Fixed in #329
